File: problems/codeforces/div1/round176/R176D1C.cpp
#include <cstdio>
#define N 1000100

int p[N],t[N],n,st[N];

int main()
{
	scanf("%d",&n);
	for(int i=0; i<n; i++)
		scanf("%d",&p[i]);
	int m;
	scanf("%d",&m);
	for(int i=0; i<m; i++)
	{
		int x;
		scanf("%d",&x);
		t[x-1]=1;
	}
	int top=-1;
	for(int i=n-1; i>=0; i--)
	{
		if(top<0||t[i]||p[i]!=st[top])
			st[++top]=p[i],p[i]=-p[i];
		else --top;
	}
	if(top==-1)
	{
		puts("YES");
		for(int i=0; i<n; i++)
			printf("%d ",p[i]);
		puts("");
	}
	else puts("NO");
	return 0;
}

File: problems/codeforces/div1/round176/r176d1c.php
<?php
function gao($n,&$p,$sg,$st) {
	$top=-1;
	for($i=$n-1; $i>=0; $i--) {
		if($top<0||$sg[$i]||$p[$i]!=$st[$top]) {
			$st[++$top]=$p[$i];
			$p[$i]=-$p[$i];
		}
		else
			--$top;
	}
	return $top==-1;
}
$n=fgets(STDIN);
$p=explode(' ',trim(fgets(STDIN)));
$q=explode(' ',trim(fgets(STDIN)));
$sg=array();
$st=array();
for($i=0; $i<$n; $i++)
	$sg[$i]=0;
for($i=1; $i<=$q[0]; $i++)
	$sg[$q[$i]-1]=1;
if(gao($n,$p,$sg,$st)) {
	echo "YES\n";
	for($i=0; $i<$n; $i++)
		echo $p[$i].' ';
	echo "\n";
}
else
	echo "NO\n";
?>
